hellabot.go

```diff
@@ -2,232 +2,175 @@ package hbot
 
 import (
 	"bufio"
-	"encoding/json"
-	"flag"
 	"fmt"
 	"net"
-	"os"
+	"sync"
 	"time"
 
+	"github.com/sorcix/irc"
+	log "gopkg.in/inconshreveable/log15.v2"
+	logext "gopkg.in/inconshreveable/log15.v2/ext"
+
 	"bytes"
 	"crypto/tls"
 	"encoding/base64"
 )
 
-// Log Levels
-const (
-	LError = iota
-	LWarning
-	LTrace
-	LNotice
-	LInfo
-	LNoise
-)
-
-// Verbosity of logging
-var Verbosity int
+// Bot implements an irc bot to be connected to a given server
+type Bot struct {
 
-type IrcCon struct {
+	// This is set if we have hijacked a connection
+	reconnecting bool
 	// Channel for user to read incoming messages
 	Incoming chan *Message
-
-	// Map of irc channels this bot is joined to
-	Channels map[string]*IrcChannel
-
-	// Channels to join after connection
-	JoinAfterConnection []string
-
-	//Server password (optional) only used if set
-	Password string
-
-	// SSL
-	UseSSL bool
-
-	// Do SASL authentication
-	DoSasl bool
-
 	con      net.Conn
 	outgoing chan string
-	tr       []*Trigger
-
-	// This bots nick
-	nick string
-
+	triggers []Trigger
+	// When did we start? Used for uptime
+	started time.Time
 	// Unix domain socket address for reconnects (linux only)
 	unixastr string
 	unixlist net.Listener
-
-	// Whether or not this is a reconnect instance
-	reconnect bool
-
+	// Log15 loggger
+	log.Logger
+	didJoinChannels sync.Once
+
+	// Exported fields
+	Host          string
+	Password      string
+	Channels      []string
+	SSL           bool
+	SASL          bool
+	HijackSession bool
+	// This bots nick
+	Nick string
 	// Duration to wait between sending of messages to avoid being
 	// kicked by the server for flooding (default 200ms)
 	ThrottleDelay time.Duration
+	// Maxmimum time between incoming data
+	PingTimeout time.Duration
 }
 
-type Config struct {
-	Server, Nick   string
-	Channels       []string
-	SSL, reconnect bool
+func (bot *Bot) String() string {
+	return fmt.Sprintf("Server: %s, Channels: %v, Nick: %s", bot.Host, bot.Channels, bot.Nick)
 }
 
-// load a Json config file
-func LoadConfig(f string) (Config, error) {
-
-	file, err := os.Open(f)
-
-	if err != nil {
-		fmt.Println("Couldn't read config file")
-		return Config{}, err
+// NewBot creates a new instance of Bot
+func NewBot(host, nick string, options ...func(*Bot)) (*Bot, error) {
```

Comment:
Look at http://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis

They are used like this

```go
hijackSession := func(bot *hbot.Bot) {                                                                                                                                                                                                                                                 
        bot.HijackSession = true                                                                                                                                                                                                                                                           
    }                                                                                                                                                                                                                                                                                      
    channels := func(bot *hbot.Bot) {                                                                                                                                                                                                                                                      
        bot.Channels = []string{"#test"}                                                                                                                                                                                                                                                   
    }                                                                                                                                                                                                                                                                                      
    irc, err := hbot.NewBot(*serv, *nick, hijackSession, channels)                                                                                                                                                                                                                         
    if err != nil {                                                                                                                                                                                                                                                                        
        panic(err)                                                                                                                                                                                                                                                                         
    }                                                               
```

You could use as many options as you wish, or you could do many changes in one option. I plan to follow the example in the link a bit closer and return an error at some point, but currently we have no options that can fail so there is no point.